src/farming/library.cairo

```diff
@@ -0,0 +1,965 @@
+    }
+
+    func prices{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        len: felt, times: felt*, prices: felt*, updated_prices: felt*, cumsales: felt*
+    ) {
+        alloc_locals;
+
+        return CarbonableFarming._compute_cumsales();
+    }
+
+    //
+    // Externals
+    //
+
+    func set_prices{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        times_len: felt, times: felt*, prices_len, prices: felt*
+    ) {
+        alloc_locals;
+
+        Array.store(key=TIME_SK, len=times_len, values=times);
+        Array.store(key=PRICE_SK, len=prices_len, values=prices);
+
+        return ();
+    }
+
+    func add_price{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        time: felt, price: felt
+    ) {
+        alloc_locals;
+
+        // [Check] Time is later than the project start time
+        let (contract_address) = CarbonableFarming_carbonable_project_address_.read();
+        let (slot) = CarbonableFarming_carbonable_project_slot_.read();
+        let (first_time) = ICarbonableProject.getStartTime(
+            contract_address=contract_address, slot=slot
+        );
+        with_attr error_message("CarbonableFarming: time is sooner than the project start time") {
+            assert_le(first_time + 1, time);
+        }
+
+        // [Check] First time to store
+        let (len) = Array.read_len(key=TIME_SK);
+        if (len == 0) {
+            // [Effect] Update storage
+            Array.add(key=TIME_SK, value=first_time);
+            Array.add(key=TIME_SK, value=time);
+            Array.add(key=PRICE_SK, value=price);
+            Array.add(key=PRICE_SK, value=price);
+
+            return ();
+        }
+
+        // [Check] Time is later than the last stored time
+        let (last_time) = Array.read(key=TIME_SK, index=len - 1);
+        with_attr error_message("CarbonableFarming: time is sooner than the last stored time") {
+            assert_le(last_time + 1, time);
+        }
+
+        // [Check] Absorption is captured between these times
+        let (initial_absorption) = ICarbonableProject.getAbsorption(
+            contract_address=contract_address, slot=slot, time=last_time
+        );
+        let (final_absorption) = ICarbonableProject.getAbsorption(
+            contract_address=contract_address, slot=slot, time=time
+        );
+        with_attr error_message("CarbonableFarming: the period must capture absorption") {
+            assert_not_zero(final_absorption - initial_absorption);
+        }
+
+        // [Effect] Update storage
+        Array.add(key=TIME_SK, value=time);
+        Array.add(key=PRICE_SK, value=price);
+
+        // [Event] Emit event
+        let (current_time) = get_block_timestamp();
+        PriceUpdate.emit(time=current_time);
+
+        return ();
+    }
+
+    func update_last_price{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        time: felt, price: felt
```

Comment:
It seems the `time` argument is not checked/used properly.
And isn't it also possible to update the last price with the same time? (seems the `assert_not_zero` would prevent it)